data/transactions/logic/eval.go

```diff
@@ -0,0 +1,710 @@
+		for i, argType := range argsTypes {
+			if !opCompat(argType, cx.stack[first+i].argType()) {
+				cx.err = fmt.Errorf("%s arg %d wanted %s but got %s", opsByOpcode[opcode].Name, i, argType.String(), cx.stack[first+i].typeName())
+			}
+		}
+	}
+	opsByOpcode[opcode].op(cx)
+	if cx.Trace != nil {
+		if len(cx.stack) == 0 {
+			fmt.Fprintf(cx.Trace, "%3d %s => %s\n", cx.pc, opsByOpcode[opcode].Name, "<empty stack>")
+		} else {
+			fmt.Fprintf(cx.Trace, "%3d %s => %s\n", cx.pc, opsByOpcode[opcode].Name, cx.stack[len(cx.stack)-1].String())
+		}
+	}
+	if cx.err != nil {
+		return
+	}
+	if len(cx.stack) > MaxStackDepth {
+		cx.err = errors.New("stack overflow")
+		return
+	}
+	if cx.nextpc != 0 {
+		cx.pc = cx.nextpc
+		cx.nextpc = 0
+	} else {
+		cx.pc++
+	}
+}
+
+func opErr(cx *evalContext) {
+	cx.err = errors.New("error")
+}
+
+func opSHA256(cx *evalContext) {
+	last := len(cx.stack) - 1
+	hash := sha256.Sum256(cx.stack[last].Bytes)
+	cx.stack[last].Bytes = hash[:]
+}
+
+// The Keccak256 variant of SHA-3 is implemented for compatibility with Ethereum
+func opKeccak256(cx *evalContext) {
+	last := len(cx.stack) - 1
+	hasher := sha3.NewLegacyKeccak256()
+	hash := hasher.Sum(cx.stack[last].Bytes)
+	cx.stack[last].Bytes = hash[:]
+}
+
+// This is the hash commonly used in Algorand in crypto/util.go Hash()
+//
+// It is explicitly implemented here in terms of the specific hash for
+// stability and portability in case the rest of Algorand ever moves
+// to a different default hash. For stability of this language, at
+// that time a new opcode should be made with the new hash.
+func opSHA512_256(cx *evalContext) {
+	last := len(cx.stack) - 1
+	hash := sha512.Sum512_256(cx.stack[last].Bytes)
+	cx.stack[last].Bytes = hash[:]
+}
+
+func opPlus(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cx.stack[prev].Uint += cx.stack[last].Uint
+	if cx.stack[prev].Uint < cx.stack[last].Uint {
+		cx.err = errors.New("+ overflowed")
+		return
+	}
+	cx.stack = cx.stack[:last]
+}
+
+func opMinus(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	if cx.stack[last].Uint > cx.stack[prev].Uint {
+		cx.err = errors.New("- would result negative")
+		return
+	}
+	cx.stack[prev].Uint -= cx.stack[last].Uint
+	cx.stack = cx.stack[:last]
+}
+
+func opDiv(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	if cx.stack[last].Uint == 0 {
+		cx.err = errors.New("/ 0")
+		return
+	}
+	cx.stack[prev].Uint /= cx.stack[last].Uint
+	cx.stack = cx.stack[:last]
+}
+
+func opModulo(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	if cx.stack[last].Uint == 0 {
+		cx.err = errors.New("% 0")
+		return
+	}
+	cx.stack[prev].Uint = cx.stack[prev].Uint % cx.stack[last].Uint
+	cx.stack = cx.stack[:last]
+}
+
+func opMul(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	a := cx.stack[prev].Uint
+	b := cx.stack[last].Uint
+	v := a * b
+	if v/a != b {
+		cx.err = errors.New("* overflowed")
```

Comment:
`v` is never assigned.  Perhaps:
```
cx.stack[prev].Uint *= cx.stack[last].Uint
if cx.stack[prev].Uint < cx.stack[last].Uint {
```

Reply:
good catch ;-)
Although I would place the assignment *after* the if clause so it would remain consistent with *no op on error"

Reply:
Sounds right.